### PR TITLE
Project setup

### PR DESCRIPTION
**Pull request by douglasmarques**
Setting up an open-source Kotlin library to hold the Prosemirror code we implemented during the Native Editor project.
I am using IntelliJ IDEA 2024.1.4 (Community Edition) to build and run the project. I am unsure if it will work with VsCode.

This is the initial commit and the repo is not public yet. I am planning to make it public after I finish some to-dos and finish the pipeline implementation to deploy the library on our public maven repository

Tasks covered in this PR:

- Copying and adapting code from MobileKit repo
- Adding license files (CREDO process - OSR-1000)
- Initial Github actions setup
- Initial maven repository publish setup (need to test)

**Update**
I needed to to make the repo public to enable Github actions

### CODE REVIEW

File: .github/workflows/gradle.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+          distribution: 'adopt'
+
+      - name: Grant execute permission for gradlew
+        run: chmod +x gradlew
+
+      - name: Build with Gradle
+        run: ./gradlew build
+
+        # Populates ARTIFACTORY_USERNAME and ARTIFACTORY_API_KEY with
+        # temporary username/password for publishing to packages.atlassian.com
```
Comment (douglasmarques) [inline .github/workflows/gradle.yml:22]: I still need to test deploy on ARTIFACTORY

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    steps:
+
+      - name: "Checkout sources"
+        uses: actions/checkout@v4
+
+      - name: "Setup Java"
+        uses: actions/setup-java@v4
+        with:
+          java-version: '8'
```
Comment (douglasmarques) [inline .github/workflows/publish.yml:26]: need to update to Java 17

File: README.md [hunk 3]
```diff
@@ -1 +1,43 @@
-# prosemirror-kotlin
+# Prosemirror Kotlin
+[![Atlassian license](https://img.shields.io/badge/license-Apache%202.0-blue.svg?style=flat-square)](LICENSE) [![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](CONTRIBUTING.md)
+
+Java/Kotlin implementation of [Prosemirror](https://prosemirror.net/)
+
+## Maven / Gradle dependency
+
+Check the latest package at Maven central on: https://mvnrepository.com/artifact/com.atlassian/prosemirror-kotlin.
```
Comment (douglasmarques) [inline README.md:8]: is not deployed yet, might need to update the packages name and versions here

File: config/ktlint/baseline.xml
Comment (douglasmarques) [inline config/ktlint/baseline.xml:1]: Need to update the Ktlint rules to be similar to our codestyle

File: renovate.json [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (douglasmarques) [inline renovate.json:1]: Need to test renovate or change to dependabot

File: src/.DS_Store
Comment (douglasmarques) [inline src/.DS_Store:1]: remove this files and add rule on gitignore

File: src/main/kotlin/com/atlassian/prosemirror/collab/Collab.kt [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    if (collabState.unconfirmed.isEmpty()) return null
+    return SendableSteps(
+        version = collabState.version,
+        steps = collabState.unconfirmed.map { it.step },
+        clientID = (collabKey.get(state) as CollabPlugin).spec.config.clientID,
+        // TODO Aleksei - what is going on here?
+        origins = emptyList()
+//        get origins() {
+//            return (this as any)._origins || ((this as any)._origins = collabState.unconfirmed.map{ s -> s.origin})
+//        }
```
Comment (douglasmarques) [inline src/main/kotlin/com/atlassian/prosemirror/collab/Collab.kt:237]: Fix this TODO

File: src/main/kotlin/com/atlassian/prosemirror/collab/collab.ts [hunk 6]
```diff
@@ -0,0 +1,184 @@
+import {Plugin, PluginKey, TextSelection, EditorState, Transaction} from "prosemirror-state"
```
Comment (douglasmarques) [inline src/main/kotlin/com/atlassian/prosemirror/collab/collab.ts:1]: Discuss with the team!
Remove ts files? Remove comments where we are copying ts code? o

File: src/main/kotlin/com/atlassian/prosemirror/model/Mark.kt [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+interface UnsupportedMark {
+    var originalMarkName: String?
+}
+
+// Indicates that text that contains that mark should not be serialised to Json when
+// sanitised = true parameter is provided
+interface JsonOmitMark
+
+@JvmInline
+value class MarkId(val id: String) : Serializable
```
Comment (douglasmarques) [inline src/main/kotlin/com/atlassian/prosemirror/model/Mark.kt:136]: Android uses parcealable, should we create a parcelable wrapper to hold these serializable objects or change android to use serialisable?

File: src/main/kotlin/com/atlassian/prosemirror/model/Mark.kt [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+                    (it as? UnsupportedMark)?.originalMarkName = jsonType
+                }
+            } else {
+                type.create(attrs).also { (it as? UnsupportedMark)?.originalMarkName = jsonType }
+            }
+        }
+
+        // Test whether two sets of marks are identical.
+        fun sameSet(a: List<Mark>, b: List<Mark>): Boolean {
+            return a == b // Aleksei: TODO confirm if it works fine
```
Comment (douglasmarques) [inline src/main/kotlin/com/atlassian/prosemirror/model/Mark.kt:240]: Create a test for it and remove comment

File: src/main/kotlin/com/atlassian/prosemirror/util/Utils.kt [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+fun getMilliseconds() = System.currentTimeMillis()
+
+fun Node.resolveAndLog(from: Int, to: Int): Pair<ResolvedPos, ResolvedPos>? {
+  try {
+    val resolvedFrom = resolve(from)
+    val resolvedTo = resolve(to)
+    return resolvedFrom to resolvedTo
+  } catch (e: RangeError) {
+    // It's safe to log Node.toString as UGC is not returned in Debug mode
```
Comment (douglasmarques) [inline src/main/kotlin/com/atlassian/prosemirror/util/Utils.kt:15]: Should find a library to log or maybe just remane this method to `resolveSafe`

File: src/test/.DS_Store
Comment (tchernykh) [inline src/test/.DS_Store:1]: Would be good to add those to ignore and remove them.

File: README.md [hunk 10]
```diff
@@ -1 +1,43 @@
...
+## Contributions
+
+Contributions to prosemirror-kotlin are welcome! Please see [CONTRIBUTING.md](CONTRIBUTING.md) for details.
+
+## License
+
+Copyright (c) 2024 Atlassian and others.
+Apache 2.0 licensed, see [LICENSE](LICENSE) file.
+
+<br/>
```
Comment (tchernykh) [inline README.md:40]: this line and some unreadable chars below does not seems correct.

File: src/main/kotlin/com/atlassian/prosemirror/model/Mark.kt (hunk 7 above)
Comment (tchernykh) [reply src/main/kotlin/com/atlassian/prosemirror/model/Mark.kt:136]: We don't want to use Android specific code here to have any chance to make this project multiplatform in the future.

File: README.md (hunk 10 above)
Comment (douglasmarques) [reply README.md:40]: There is a template to add it, but renders ok:

image

File: src/main/kotlin/com/atlassian/prosemirror/model/Mark.kt (hunk 7 above)
Comment (douglasmarques) [reply src/main/kotlin/com/atlassian/prosemirror/model/Mark.kt:136]: Correct, I mean to create a wrapper in MK

**Review by tchernykh — APPROVED** [review]
Looks good.

**Pull request merged**